pkg/objstore/objstore.go

```diff
@@ -437,6 +437,87 @@ func (b *metricBucket) Name() string {
 	return b.bkt.Name()
 }
 
+type prefixedBucket struct {
+	bkt    Bucket
+	prefix string
+}
+
+// NewPrefixedBucket returns a new prefixedBucket.
+func NewPrefixedBucket(bkt Bucket, prefix string) *prefixedBucket {
+	return &prefixedBucket{
+		bkt:    bkt,
+		prefix: prefix,
+	}
+}
+
+func (pbkt *prefixedBucket) WithExpectedErrs(expectedFunc IsOpFailureExpectedFunc) Bucket {
+	if ib, ok := pbkt.bkt.(InstrumentedBucket); ok {
+		return &prefixedBucket{
+			bkt:    ib.WithExpectedErrs(expectedFunc),
+			prefix: pbkt.prefix,
+		}
+	}
+	return pbkt
+}
+
+func (pbkt *prefixedBucket) ReaderWithExpectedErrs(expectedFunc IsOpFailureExpectedFunc) BucketReader {
+	return pbkt.WithExpectedErrs(expectedFunc)
+}
+
+func (pbkt *prefixedBucket) Name() string {
+	return pbkt.bkt.Name()
+}
+
+func (pbkt *prefixedBucket) Upload(ctx context.Context, name string, r io.Reader) (err error) {
+	pname := filepath.Join(pbkt.prefix, name)
+	err = pbkt.bkt.Upload(ctx, pname, r)
+	return
+}
+
+func (pbkt *prefixedBucket) Delete(ctx context.Context, name string) (err error) {
+	pname := filepath.Join(pbkt.prefix, name)
```

Comment:
You should juse `path.Join()` and not `filepath.Join()`. The former guarantees paths are joined with `/`, which is what object stores expect.

_Same comment applies across the change set._

Reply:
Aso the `.Join()` is used quite a lot. You could add an help `prefixedBucket.nameWithPrefix(name string)`

Reply:
yeah, might be good to add a helper function but I don't know how will it reduce the redundancy of the code, as I'll have to call the helper function same number of times as I have called filepath.Join().
But I'll update the code with your suggestion. I think it will add consistency to the code.